### PR TITLE
feat(readme): Add packages/common/README.md

### PR DESCRIPTION
**Pull request by jbrundage**
Add 10 common samples

Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith I wasn't quite sure which exported common widgets to showcase. I'm particularly excited about the SVGWidget/HTMLWidget/CanvasWidget samples. I'll use them as boilerplate code. Eating this "dog food" has me thinking about the possibility of a WebglWidget heh.

File: demos/gallery/samples/common/CanvasWidget.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        this._ctx.fillText(row[0], gutter, (rowHeight * i) + (gutter * (i+1)) + (rowHeight/2));
+        this._ctx.fillRect(
+            gutter + labelWidth + gutter, 
+            (rowHeight * i) + (gutter * (i+1)),
+            row[1], 
+            rowHeight - gutter
+        );
+    })
+};
+w.render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/CanvasWidget.js]: Change this to be a new `class` which derived from CanvasWidget with that update code:
```
class MyCanvasWidget extends CanvasWidget {
...
}
const w = new MyCanvasEidget().target(...)
```
Type of thing

File: demos/gallery/samples/common/HTMLWidget.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            rect.style.height = rowHeight+"px";
+        })
+        .attr("transform",(row,i)=>{
+            const x = gutter;
+            const y = (rowHeight * i) + (gutter * (i+1)) + rowHeight;
+            return `translate(${x} ${y})`;
+        })
+        ;
+};
+w.render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Add derived class similar to CanvasWidget

File: demos/gallery/samples/common/SVGWidget.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        })
+        .attr("width",(row,i)=>{
+            return row[1];
+        })
+        .attr("height",(row,i)=>{
+            return rowHeight;
+        })
+        ;
+};
+w.render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/SVGWidget.js]: Derived class also

File: demos/gallery/samples/common/TitleBar.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+Button.prototype.publish("faClass", "", "string", "FontAwesome class name (if done right will look like fa-user)");
+Button.prototype.publish("label", "", "string", "The label/text written on the button (to the right of the faChar (if a faChar is provided))");
+Button.prototype.publish("lastFAChar", "", "string", "The 'last-icon' displays on the far right face of the button. FontAwesome character (if done right will look like a  in your favorite editor)");
+Button.prototype.publish("lastFAClass", "", "string", "The label/text written on the button (to the left of the 'last-icon' (if a faChar is provided)). The 'last-icon' displays on the far right face of the button.");
+Button.prototype.publish("tooltip", "", "string", "Displays as the button alt text attribute");
+Button.prototype.publish("state", "enabled", "set", "Current 'state' of the button is usually 'enabled'.", ["enabled", "disabled", "selected"]);
+Button.prototype.publish("styleMode", "default", "set", "Several different base button styles.", ["default", "Bootstrap1", "Bootstrap2", "Bootstrap3", "Bootstrap4", "Primer"]);
+*/
```
Comment (GordonSmith) [inline demos/gallery/samples/common/TitleBar.js]: Remove commented code (you could add a comment with a link to the button sources if you thought it was a good idea?)

File: packages/common/src/TitleBar.ts [hunk 5]
```diff
@@ -59,11 +59,26 @@ Button.prototype._class += " common_Button";
...
+Button.prototype.publish("faClass", "", "string", "FontAwesome class name (if done right will look like fa-user)");
+Button.prototype.publish("label", "", "string", "The label/text written on the button (to the right of the faChar (if a faChar is provided))");
+Button.prototype.publish("lastFAChar", "", "string", "The 'last-icon' displays on the far right face of the button. FontAwesome character (if done right will look like a  in your favorite editor)");
+Button.prototype.publish("lastFAClass", "", "string", "The label/text written on the button (to the left of the 'last-icon' (if a faChar is provided)). The 'last-icon' displays on the far right face of the button.");
+Button.prototype.publish("tooltip", "", "string", "Displays as the button alt text attribute");
+Button.prototype.publish("state", "enabled", "set", "Current 'state' of the button is usually 'enabled'.", ["enabled", "disabled", "selected"]);
+Button.prototype.publish("styleMode", "default", "set", "Several different base button styles.", ["default", "Bootstrap1", "Bootstrap2", "Bootstrap3", "Bootstrap4", "Primer"]);
```
Comment (GordonSmith) [inline packages/common/src/TitleBar.ts]: I don't think this code belongs here?

File: demos/gallery/samples/common/CanvasWidget.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            this._ctx.fillRect(
+                gutter + labelWidth + gutter, 
+                (rowHeight * i) + (gutter * (i+1)),
+                row[1], 
+                rowHeight - gutter
+            );
+        })
+    }
+}
+new CanvasExample().target("target").data(data).render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/CanvasWidget.js]: Break into several lines like all the other examples.

File: demos/gallery/samples/common/CanvasWidget.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    ["A", 100],
+    ["B", 162],
+    ["C", 262],
+];
+class CanvasExample extends CanvasWidget{
+    constructor(){
+        super();
+    }
+    enter(domNode, element){
+        super.enter.apply(this, arguments);
```
Comment (GordonSmith) [inline demos/gallery/samples/common/CanvasWidget.js]: change to `super.enter(domNode, element)` and do the same of all siimlar examples

File: demos/gallery/samples/common/EntityVertex.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    .title("EntityVertex Title")
+    .iconColor("#2c3e50")
+    .iconColorFill("#ecf0f1")
+    .iconColorStroke("#2c3e50")
+    .textboxColorFill("#ecf0f1")
+    .textboxColorStroke("#2c3e50")
+    .annotationIcons([
+        { faChar: "A", image_colorFill: "#2c3e50", shape_colorFill: "#f1c40f", shape_colorStroke: "#2c3e50" },
+        { faChar: "B", image_colorFill: "#2c3e50", shape_colorFill: "#e67e22", shape_colorStroke: "#2c3e50" },
+        { faChar: "C", image_colorFill: "#2c3e50", shape_colorFill: "#e74c3c", shape_colorStroke: "#2c3e50" },
```
Comment (GordonSmith) [inline demos/gallery/samples/common/EntityVertex.js]: trailing comma

File: demos/gallery/samples/common/EntityRect.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    .titleFontSize(18)
+    .titleColor("#ecf0f1")
+    .iconColor("#ecf0f1")
+    .backgroundShape("rect")
+    .backgroundColorFill("#2980b9")
+    .backgroundColorStroke("#2c3e50")
+    .annotationIcons([
+        { faChar: "A", image_colorFill: "#2c3e50", shape_colorFill: "#f1c40f", shape_colorStroke: "none" },
+        { faChar: "B", image_colorFill: "#2c3e50", shape_colorFill: "#e67e22", shape_colorStroke: "none" },
+        { faChar: "C", image_colorFill: "#2c3e50", shape_colorFill: "#e74c3c", shape_colorStroke: "none" },
```
Comment (GordonSmith) [inline demos/gallery/samples/common/EntityRect.js]: remove trailing comma

File: demos/gallery/samples/common/EntityPin.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    .backgroundShape("pin")
+    .backgroundColorFill("#F8F8F8")
+    .backgroundColorStroke("#CCCCCC")
+    .cornerRadius(5)
+    .arrowHeight(10)
+    .arrowWidth(16)
+    .annotationIcons([
+        {faChar: "A",image_colorFill: "#E3151A",shape_colorFill: "none",shape_colorStroke: "#E3151A",shape: "circle"},
+        {faChar: "B",image_colorFill: "#E3151A",shape_colorFill: "none",shape_colorStroke: "#E3151A",shape: "circle"},
+        {faChar: "C",image_colorFill: "#E3151A",shape_colorFill: "none",shape_colorStroke: "#E3151A",shape: "circle"},
```
Comment (GordonSmith) [inline demos/gallery/samples/common/EntityPin.js]: remove trailing comma

File: demos/gallery/samples/common/HTMLWidget.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+                rect.style.backgroundColor = "red";
+                rect.style.width = row[1]+"px";
+                rect.style.height = rowHeight+"px";
+            })
+            .attr("transform",(row,i)=>{
+                const x = gutter;
+                const y = (rowHeight * i) + (gutter * (i+1)) + rowHeight;
+                return `translate(${x} ${y})`;
+            })
+            ;
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Can you clean up the `d3` code so its more "standard d3".

Missing `exit` call

File: demos/gallery/samples/common/SVGWidget.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                const y = (rowHeight * i) + (gutter * (i+1)) + (rowHeight/2);
+                return `translate(${x - (w/2)} ${y - (h/2)})`;
+            })
+            .attr("width",(row,i)=>{
+                return row[1];
+            })
+            .attr("height",(row,i)=>{
+                return rowHeight;
+            })
+            ;
```
Comment (GordonSmith) [inline demos/gallery/samples/common/SVGWidget.js]: MIssing `exit` code

File: demos/gallery/samples/common/HTMLWidget.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    }
+}
+new HTMLExample().target("target").data(data).render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: split into several lines

File: demos/gallery/samples/common/SVGWidget.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+new SVGExample().target("target").data(data).render();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/SVGWidget.js]: split into several lines

File: packages/common/src/CanvasWidget.ts [hunk 15]
```diff
@@ -11,9 +12,14 @@ export class CanvasWidget extends HTMLWidget {
         const retVal = super.resize(size);
         this._element.attr("width", this._size.width);
         this._element.attr("height", this._size.height);
+        this.resetContext();
         return retVal;
     }
 
+    resetContext() {
+        this._ctx = this._element.node().getContext("2d");
+    }
```
Comment (GordonSmith) [inline packages/common/src/CanvasWidget.ts]: I don't think this can work as there has been no `target` assigned?

File: packages/common/src/CanvasWidget.ts [hunk 16]
```diff
@@ -11,9 +12,14 @@ export class CanvasWidget extends HTMLWidget {
         const retVal = super.resize(size);
         this._element.attr("width", this._size.width);
         this._element.attr("height", this._size.height);
```
Comment (GordonSmith) [inline packages/common/src/CanvasWidget.ts]: I don't think this can work as there has been no `target` assigned?

Comment (jbrundage) [reply packages/common/src/CanvasWidget.ts]: `resize` isn't called until after there's been a target assigned

**Comment by jbrundage** [comment]
@GordonSmith I've made the requested changes and added the actual README.md

File: demos/gallery/samples/common/HTMLWidget.js (hunk 11 above)
Comment (GordonSmith) [reply demos/gallery/samples/common/HTMLWidget.js]: @jbrundage bump

File: packages/common/src/CanvasWidget.ts [hunk 17]
```diff
@@ -2,18 +2,31 @@ import { HTMLWidget } from "./HTMLWidget";
...
 export class CanvasWidget extends HTMLWidget {
 
+    _ctx: CanvasRenderingContext2D;
     constructor() {
         super();
         this._tag = "canvas";
     }
 
     resize(size) {
+        console.log("resize");
```
Comment (GordonSmith) [inline packages/common/src/CanvasWidget.ts]: Remove console.log

File: packages/common/src/CanvasWidget.ts [hunk 18]
```diff
@@ -2,18 +2,31 @@ import { HTMLWidget } from "./HTMLWidget";
...
         const retVal = super.resize(size);
-        this._element.attr("width", this._size.width);
-        this._element.attr("height", this._size.height);
+        try {
+            console.log("this._element", this._element);
+            console.log("this._element.node()", this._element.node());
```
Comment (GordonSmith) [inline packages/common/src/CanvasWidget.ts]: remove console.log

File: packages/common/src/CanvasWidget.ts (hunk 17 above)
Comment (GordonSmith) [reply packages/common/src/CanvasWidget.ts]: @jbrundage bump

File: packages/common/src/CanvasWidget.ts (hunk 18 above)
Comment (GordonSmith) [reply packages/common/src/CanvasWidget.ts]: @jbrundage bump

File: demos/gallery/samples/common/CanvasWidget.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    ["B", 162],
+    ["C", 262]
+];
+class CanvasExample extends CanvasWidget{
+    constructor(){
+        super();
+    }
+    enter(domNode, element){
+        super.enter(domNode, element);
+        this.resize();
```
Comment (GordonSmith) [inline demos/gallery/samples/common/CanvasWidget.js]: Why is this needed?  Should it be in the CanvasWIdget?

File: demos/gallery/samples/common/HTMLWidget.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        const rowsEnter = this._div.selectAll(".data-row").data(this.data())
+            .enter()
+            .append("div")
+            .style("width","100%")
+            .style("height",rowHeight+"px")
+            .style("margin-top",gutter+"px")
+            .classed("data-row", true)
+            ;
+        rowsEnter.append("span");
+        rowsEnter.append("div");
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: I typicaly put these inside an "each" callback.

File: demos/gallery/samples/common/HTMLWidget.js [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+        rowsEnter
+            .merge(rowsEnter)
+            .style("color","red")
+            .style("float","left")
+            .each(function(row,i){
+                const text = this.children[0];
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Change to `d3Select("span")`

File: demos/gallery/samples/common/HTMLWidget.js [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+                text.style.float = "left";
+                text.style.display = "inline-block";
+                text.style.width = labelWidth+"px";
+                text.innerText = row[0];
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Use d3 paradigm for setting attrs.

File: demos/gallery/samples/common/HTMLWidget.js [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+                const rect = this.children[1];
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Change to `d3Select("div")`

File: demos/gallery/samples/common/HTMLWidget.js [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+                rect.style.float = "left";
+                text.style.display = "inline-block";
+                rect.style.marginLeft = gutter+"px";
+                rect.style.backgroundColor = "red";
+                rect.style.width = row[1]+"px";
+                rect.style.height = rowHeight+"px";
```
Comment (GordonSmith) [inline demos/gallery/samples/common/HTMLWidget.js]: Use d3 paradigm for setting styles

File: demos/gallery/samples/common/HTMLWidget.js (hunk 11 above)
Comment (GordonSmith) [reply demos/gallery/samples/common/HTMLWidget.js]: Missing exit call.

File: demos/gallery/samples/common/SVGWidget.js [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    ["B", 162],
+    ["C", 262]
+];
+class SVGExample extends SVGWidget{
+    constructor(){
+        super();
+    }
+    enter(domNode, element){
+        super.enter(domNode, element);
+        this._g = element.append("g");
```
Comment (GordonSmith) [inline demos/gallery/samples/common/SVGWidget.js]: `_g` has not been declared

File: demos/gallery/samples/common/SVGWidget.js [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+            .classed("data-text", true)
+            .merge(texts)
+            .attr("fill","red")
+            .attr("transform",(row,i)=>{
+                const x = gutter;
+                const y = (rowHeight * i) + (gutter * (i+1)) + rowHeight;
+                return `translate(${x - (w/2)} ${y - (h/2)})`;
+            })
+            .text(row=>row[0])
+            ;
```
Comment (GordonSmith) [inline demos/gallery/samples/common/SVGWidget.js]: Missing exit call.

**Comment by GordonSmith** [comment]
@jbrundage bump

File: demos/gallery/samples/common/HTMLWidget.js (hunk 20 above)
Comment (jbrundage) [reply demos/gallery/samples/common/HTMLWidget.js]: IMO this implementation is more abstract and this is just sample code.

File: demos/gallery/samples/common/SVGWidget.js (hunk 25 above)
Comment (jbrundage) [reply demos/gallery/samples/common/SVGWidget.js]: That line is the `_g` declaration

File: demos/gallery/samples/common/CanvasWidget.js (hunk 19 above)
Comment (jbrundage) [reply demos/gallery/samples/common/CanvasWidget.js]: Hmm... evidently it isn't needed after all

**Comment by jbrundage** [comment]
@GordonSmith the requested changes have been made
